lib/api/core/storage/models/apiKey.js

```diff
@@ -0,0 +1,141 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const
+  errorsManager = require('../../../../util/errors'),
+  BaseModel = require('./baseModel');
+class ApiKey extends BaseModel {
+  constructor (_source, _id = null) {
+    super(_source, _id);
+  }
+
+  /**
+   * @override
+   */
+  async _afterDelete () {
+    const token = await BaseModel.kuzzle.repositories.token.verifyToken(
+      this.token);
+
+    await BaseModel.kuzzle.repositories.token.expire(token);
+  }
+
+  serialize ({ includeToken=false } = {}) {
+    const serialized = super.serialize();
+
+    if (! includeToken) {
+      delete serialized._source.token;
+    }
+
+    return serialized;
+  }
+
+  // Static public methods =====================================================
+
+  /**
+   * @override
+   */
+  static get collection () {
+    return 'api-keys';
+  }
+
+  /**
+   * @override
+   */
+  static get fields () {
+    return ['userId', 'hash', 'description', 'expiresAt', 'ttl', 'token'];
+  }
+
+  /**
+   * Creates a new API key for an user
+   *
+   * @param {User} user
+   * @param {String} connectionId
+   * @param {String} expiresIn - API key expiration date in ms format
+   * @param {String} description
+   * @param {Object} options - creatorId (null), apiKeyId (null), refresh (null)
+   *
+   * @returns {ApiKey}
+   */
+  static async create (
+    user,
+    connectionId,
+    expiresIn,
+    description,
+    { creatorId=null, apiKeyId=null, refresh } = {}
+  ) {
+    const
+      token = await BaseModel.kuzzle.repositories.token.generateToken(
+        user,
+        connectionId,
+        { expiresIn, bypassMaxTTL: true }),
+      hash = BaseModel.kuzzle.constructor.hash(token.jwt),
+      apiKey = new ApiKey({
+        hash,
+        description,
+        userId: user._id,
+        expiresAt: token.expiresAt,
+        ttl: token.ttl,
+        token: token.jwt
+      }, apiKeyId);
+
+    await apiKey.save({ userId: creatorId, refresh });
+
+    return apiKey;
+  }
+
+  /**
+   * Loads an user API key from the database
+   *
+   * @param {String} userId - User ID
+   * @param {String} id - API key ID
+   *
+   * @returns {ApiKey}
+   */
+  static async load (userId, id) {
+    const apiKey = await super.load(id);
+
+    if (userId !== apiKey.userId) {
+      errorsManager.throw('services', 'storage', 'not_found', id, {
+        message: `ApiKey "${id}" not found for user "${userId}".`
+      });
+    }
+
+    return apiKey;
+  }
+
+  /**
+   * Deletes API keys for an user
+   *
+   * @param {User} user
+   * @param {Object} options - refresh (null)
+   *
+   * @returns {Promise}
+   */
+  static async deleteByUser (user, { refresh } = {}) {
+    return this.deleteByQuery({ term: { userId: user._id } }, { refresh });
+  }
+}
+
+BaseModel.register(ApiKey);
```

Comment:
I don' get why we need a static register method. Can't we simply inherit from BaseModel?

Reply:
I need to define the getters and setters for the new model. I didn't find another way to not have to duplicate the code.